File: Problems/1547. (H) Minimum Cost to Cut a Stick/Python3/solution.py
class Solution:
  def minCost(self, n: int, cuts: List[int]) -> int:
    # bottom-up dynamic programming (tabulation)

    m = len(cuts)
    cuts.sort()
    cuts.insert(0, 0)
    cuts.append(n)
    # dp[i][j] is minimum cost for all cuts between i and j
    dp = [[0] * (m+2) for _ in range(m+2)]
    
    for i in range(2, m+2):
      for l in range(m+2-i):
        r = l + i
        ans = float('inf')
        for j in range(l+1, r):
          ans = min(ans, dp[l][j]+dp[j][r]+cuts[r]-cuts[l])
        dp[l][r] = ans
    
    return dp[0][-1]

File: Problems/1547. (H) Minimum Cost to Cut a Stick/Java/solution.java
class Solution {
  public int minCost(int n, int[] cuts) {
    // bottom-up dynamic programming (tabulation)

    int m = cuts.length;
    Arrays.sort(cuts);
    int[] cuts_new = new int[m+2];
    for (int i = 0; i < m+2; i++) {
      if (i == 0)
        cuts_new[i] = 0;
      else if (i == m+1)
        cuts_new[i] = n;
      else
        cuts_new[i] = cuts[i-1];
    }
    int[][] dp = new int[m+2][m+2];

    for (int i = 2; i < m+2; i++) {
      for (int l = 0; l < m+2-i; l++) {
        int r = l + i;
        int ans = Integer.MAX_VALUE;
        for (int j = l+1; j < r; j++)
          ans = Math.min(ans, dp[l][j]+dp[j][r]+cuts_new[r]-cuts_new[l]);
        dp[l][r] = ans;
      }
    }
    return dp[0][m+1];
  }
}
